deps/rabbit/src/rabbit_binding.erl

```diff
@@ -710,76 +375,56 @@ merge_entry({X1, Deleted1, Bindings1}, {X2, Deleted2, Bindings2}) ->
-                         Bs = lists:flatten(Bindings),
-                         x_callback(transaction, X, delete, Bs),
-                         {X, deleted, Bs, none};
-                     (_XName, {X, not_deleted, Bindings}) ->
-                         Bs = lists:flatten(Bindings),
-                         x_callback(transaction, X, remove_bindings, Bs),
-                         {X, not_deleted, Bs, rabbit_exchange:serial(X)}
-                 end, Deletions),
-    fun() ->
-            dict:fold(fun (XName, {X, deleted, Bs, Serial}, ok) ->
-                              ok = rabbit_event:notify(
-                                     exchange_deleted,
-                                     [{name, XName},
-                                      {user_who_performed_action, ActingUser}]),
-                              del_notify(Bs, ActingUser),
-                              x_callback(Serial, X, delete, Bs);
-                          (_XName, {X, not_deleted, Bs, Serial}, ok) ->
-                              del_notify(Bs, ActingUser),
-                              x_callback(Serial, X, remove_bindings, Bs)
-                      end, ok, AugmentedDeletions)
-    end.
+notify_deletions({error, not_found}, _) ->
+    ok;
+notify_deletions(Deletions, ActingUser) ->
+    dict:fold(fun (XName, {_X, deleted, Bs, _}, ok) ->
+                      notify_exchange_deletion(XName, ActingUser),
+                      notify_bindings_deletion(Bs, ActingUser);
+                  (_XName, {_X, not_deleted, Bs, _}, ok) ->
+                      notify_bindings_deletion(Bs, ActingUser);
+                  (XName, {_X, deleted, Bs}, ok) ->
+                      notify_exchange_deletion(XName, ActingUser),
+                      notify_bindings_deletion(Bs, ActingUser);
+                  (_XName, {_X, not_deleted, Bs}, ok) ->
+                      notify_bindings_deletion(Bs, ActingUser)
+              end, ok, Deletions).
+
+notify_exchange_deletion(XName, ActingUser) ->
+    ok = rabbit_event:notify(
+           exchange_deleted,
+           [{name, XName},
+            {user_who_performed_action, ActingUser}]).
+
+notify_bindings_deletion(Bs, ActingUser) ->
+    [rabbit_event:notify(binding_deleted,
+                         info(B) ++ [{user_who_performed_action, ActingUser}])
+     || B <- Bs],
+    ok.
 
-del_notify(Bs, ActingUser) -> [rabbit_event:notify(
-                               binding_deleted,
-                               info(B) ++ [{user_who_performed_action, ActingUser}])
-                             || B <- Bs],
-                              ok.
-
-x_callback(Serial, X, F, Bs) ->
-    ok = rabbit_exchange:callback(X, F, Serial, [X, Bs]).
-
--spec populate_index_route_table() -> ok.
-populate_index_route_table() ->
-    rabbit_misc:execute_mnesia_transaction(
-      fun () ->
-              mnesia:read_lock_table(rabbit_route),
-              mnesia:read_lock_table(rabbit_exchange),
```

Comment:
This lock of `rabbit_exchange` is missing in the moved function in `rabbit_db_binding`. Is it on purpose?

Reply:
They're there: https://github.com/rabbitmq/rabbitmq-server/pull/6821/files/f93b5dccc9909f96933eff11cf7bc81f0fc67dc3#diff-ba4aec2563f03086e68cfd65539cabb6008cb8ad20904c9802830ee67e2eb6dfR482

Reply:
This one: https://github.com/rabbitmq/rabbitmq-server/pull/6821/files#diff-ba4aec2563f03086e68cfd65539cabb6008cb8ad20904c9802830ee67e2eb6dfR499

Reply:
Done